# Spam Detector

In [28]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [29]:
# Import the data
#datapath = 
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")

data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

In [ ]:
#I think the Logistic Regression will peform better because the nature of the question is binary and that this is the fastest most efficent method.

## Split the Data into Training and Testing Sets

In [30]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"]
X = data.copy()
X = data.drop(columns="spam")


In [31]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [32]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [33]:
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
2945            0.00               0.00           0.00           0.0   
4215            0.00               0.00           0.39           0.0   
591             0.45               0.68           0.68           0.0   
4095            0.00               0.00           0.00           0.0   
1142            0.00               0.00           1.26           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
2945           0.37            0.00              0.00                0.00   
4215           0.39            0.39              0.00                0.00   
591            1.92            0.00              0.56                0.45   
4095           0.00            0.00              0.00                0.00   
1142           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
2945              0.0            0.00  ...                   0.0        0.000   
4215              0.0            0.00  ...                   0.0        0.000   
591               0.0            0.45  ...                   0.0        0.019   
4095              0.0            0.00  ...                   0.0        0.000   
1142              0.0            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
2945        0.054        0.000        0.000        0.054        0.000   
4215        0.065        0.065        0.261        0.065        0.000   
591         0.057        0.000        0.574        0.134        0.019   
4095        0.743        0.000        0.371        0.000        0.371   
1142        0.000        0.000        1.900        0.475        0.000   

      capital_run_length_average  capital_run_length_longest  \
2945                       1.066                           4   
4215                       2.890                          55   
591                        3.155                          94   
4095                       1.714                          11   
1142                       1.763                          11   

      capital_run_length_total  
2945                       128  
4215                       159  
591                        385  
4095                        24  
1142                        67  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [35]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instances
scaler = StandardScaler()

In [36]:
# Fit the Standard Scaler with the training data
X_scaler = scaler.fit(X_train)

In [38]:
# Scale the training data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [59]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression()
lr_model.fit(X_train_scaled,y_train)
print(f"Training Data Score:  {lr_model.score(X_train_scaled,y_train)}")
print(f"Testing Data Score:  {lr_model.score(X_test_scaled,y_test)}")

Training Data Score:  0.9330434782608695
Testing Data Score:  0.9235447437011295


In [60]:
# Make and save testing predictions with the saved logistic regression model using the test data
predictions = lr_model.predict(X_test_scaled)
# Review the predictions
predictions

array([0, 0, 0, ..., 0, 0, 0])

In [61]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test,predictions)

0.9235447437011295

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [62]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
random_model = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_scaled,y_train)

print(f'Training Score: {random_model.score(X_train_scaled, y_train)}')
print(f'Testing Score: {random_model.score(X_test_scaled, y_test)}')

Training Score: 0.9997101449275362
Testing Score: 0.9409209383145091


In [63]:
# Make and save testing predictions with the saved logistic regression model using the test data
predictions2 = random_model.predict(X_test_scaled)

# Review the predictions
predictions2

array([0, 0, 1, ..., 0, 0, 0])

In [64]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test,predictions2)
acc_score

0.9409209383145091

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

#The random forest model performed better for the analysis - yielding a 94% compared to a 92% for the logistic regression.  The Random Forest model was able to group and define more like elements within the data set to better predict where the mail item should go - spam or inbox.